### PR TITLE
[SPARK-49476][SQL] Fix nullability of base64 function

### PR DESCRIPTION
**Pull request by Kimahriman**
### What changes were proposed in this pull request?

Fix the nullability of the `Base64` expression to be based on the child's nullability, and not always be nullable.

### Why are the changes needed?

https://github.com/apache/spark/pull/47303 had a side effect of changing the nullability by the switch to using `StaticInvoke`. This was also backported to Spark 3.5.2 and caused schema mismatch errors for stateful streams when we upgraded. This restores the previous behavior which is supported by StaticInvoke through the `returnNullable` argument. If the child is non-nullable, we know the result will be non-nullable.

### Does this PR introduce _any_ user-facing change?

Restores the nullability of the `Base64` expression to what is was in Spark 3.5.1 and earlier.

### How was this patch tested?

New UT

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by Kimahriman** [comment]
@dongjoon-hyun @yaooqinn @wForget

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala [hunk 1]
```diff
@@ -467,6 +467,12 @@ class StringExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
     val b = $"b".binary.at(0)
     val bytes = Array[Byte](1, 2, 3, 4)
 
+    assert(!Base64(Literal(bytes)).nullable)
+    assert(Base64(Literal.create(null, BinaryType)).nullable)
+    assert(!UnBase64(Literal("AQIDBA==")).nullable)
+    assert(UnBase64(Literal.create(null, StringType)).nullable)
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala:473]: Let's add an check when you pass not-NULL expr but mark it as a `nullable`:
```scala
assert(Base64(Literal(bytes).castNullable()).nullable)
```

**Review by MaxGekk — COMMENTED** [review]
LGTM except of a comment.

Please, append the `[SQL]` tag in PR's title.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala (hunk 1 above)
Comment (Kimahriman) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala]: Added

**Review by yaooqinn — APPROVED** [review]

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala [hunk 2]
```diff
@@ -467,6 +467,14 @@ class StringExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
...
     val bytes = Array[Byte](1, 2, 3, 4)
 
+    assert(!Base64(Literal(bytes)).nullable)
+    assert(Base64(Literal.create(null, BinaryType)).nullable)
+    assert(Base64(Literal(bytes).castNullable()).nullable)
+    assert(!UnBase64(Literal("AQIDBA==")).nullable)
+    assert(UnBase64(Literal.create(null, StringType)).nullable)
+    assert(UnBase64(Literal("AQIDBA==").castNullable()).nullable)
+
+
```
Comment (LuciferYang) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala:477]: nit: Adding one blank line here is enough

**Review by LuciferYang — APPROVED** [review]

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala (hunk 2 above)
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala:477]:
```suggestion
```

**Review by MaxGekk — APPROVED** [review]
+1, LGTM. Merging to master/3.5.
Thank you, @Kimahriman and @LuciferYang @yaooqinn for review.

**Comment by Kimahriman** [comment]
Made a follow up to fix a test in the 3.5 backport https://github.com/apache/spark/pull/47964

**Comment by dongjoon-hyun** [comment]
Thank you, @Kimahriman and all.

**Pull request closed without merging**